### PR TITLE
Index expression

### LINKED ISSUE #41: Support field access with []
Currently, to get the value of a field, one can use the `.get()` function defined on the object prototype.
Similar to other languages, in particular js, this should be also supported using square brackets:
e.g. `test["value"]` is equivalent to `test.value` is equivalent to `test.get("value")`

main reasons are
- similar to js, feature users might expect, in particular for numerical keys
- simplifies field access (less code to write, more concise)
- why not (no use for the square brackets yet in this context)

autocomplete support should also be adapted to fall back to this syntax when numerical keys or strings with spaces are used

### PR DESCRIPTION
**Pull request by nk-coding**
## technical changes
- add support for index expressions (like in e.g. js, python, kotlin, ...)
  - access: `this["my expression"]`
  - invoke: `this["my expression"]()`
  - set: `this["my expression"] = true`

## new features
- index expressions
- completion: when using dot syntax and autocompleting an item which cannot be used with the dot syntax, automatically switch to the index syntax
- index expressions in class diagram property/method syntax to add multiplicity (like in UML standard)
  - `test : int [0 .. *]`

## issues
- closes #41

### CODE REVIEW

File: packages/core/src/parser/cstToAstTransformer.ts [hunk 1]
```diff
@@ -276,17 +268,70 @@ export function generateCstToAstTransfromer(parser: Parser): ICstVisitor<never,
...
+            let identifier: IToken | undefined = undefined;
+            let index: Expression | undefined = undefined;
+            if (ctx.expression) {
+                index = this.visit(ctx.expression[0]);
+            } else {
+                identifier = ctx.Identifier[0];
+            }
+            return (baseExpression) => {
+                const startPos = baseExpression.range;
+                if (identifier != undefined) {
```
Comment (delvh) [inline packages/core/src/parser/cstToAstTransformer.ts:289]: Method is too long and does too much.
Extract submethods.

File: packages/diagram/src/module/diagrams/classDiagramModule.ts [hunk 2]
```diff
@@ -45,6 +47,56 @@ function convertStringOrIdentifier(expression: Expression): string {
...
+    }
+}
+
+/**
+ * Converts a multiplicity element, can either be a string or a range using the '..' operator
+ *
+ * @param expression the expression to convert
+ * @returns the multiplicity element without enclosing square brackets
+ */
+function convertMultiplicityElement(expression: Expression): string {
```
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/diagram/src/module/diagrams/classDiagramModule.ts:73]: probably, in particular for the component diagram
for that, we probably need to refactor the whole UML stuff a lot
especially as some also want a general `umlDiagram` type which allows to use elements of different UML diagram types

File: packages/core/src/parser/cstToAstTransformer.ts (hunk 1 above)
Comment (nk-coding) [reply packages/core/src/parser/cstToAstTransformer.ts:288]: I knew I forgot sth
c0f53c49eac7c2f07875ca478d4e9f3647a3502d

File: packages/diagram/src/module/diagrams/classDiagramModule.ts [hunk 3]
```diff
@@ -45,6 +47,56 @@ function convertStringOrIdentifier(expression: Expression): string {
...
+ * @param expression the expression to convert
+ * @returns the string representation of the expression
+ */
+function convertNumberOrIdentifier(expression: Expression): string {
+    if (expression instanceof NumberLiteralExpression) {
+        return expression.value.toString();
+    } else if (expression instanceof IdentifierExpression) {
+        return expression.identifier;
+    } else {
+        throw new RuntimeError("Expected number or identifier");
```
[the remark this answers is not in the document]
Comment (nk-coding) [reply packages/diagram/src/module/diagrams/classDiagramModule.ts]: changed with 118852a63b0ea01cb174d4d37da2284900fb1715
also changed the other error message

**Review by delvh — APPROVED** [review]

**Pull request merged**